File: src/services/vehicleService.ts
import apiClient from './api';
import { Vehicle, VehicleRentalStatus } from '../types';

export const vehicleService = {
  async getAll(): Promise<Vehicle[]> {
    const { data } = await apiClient.get<Vehicle[]>('/vehicles');
    return data;
  },

  async getById(id: string): Promise<Vehicle> {
    const { data } = await apiClient.get<Vehicle>(`/vehicles/${id}`);
    return data;
  },

  async getByDestination(destinationId: string): Promise<Vehicle[]> {
    const { data } = await apiClient.get<Vehicle[]>(`/vehicles/destination/${destinationId}`);
    return data;
  },

  async getAvailable(): Promise<Vehicle[]> {
    const { data } = await apiClient.get<Vehicle[]>('/vehicles/available');
    return data;
  },

  async create(vehicle: Omit<Vehicle, 'id'>): Promise<Vehicle> {
    const { data } = await apiClient.post<Vehicle>('/vehicles', vehicle);
    return data;
  },

  async update(id: string, vehicle: Partial<Vehicle>): Promise<Vehicle> {
    const { data } = await apiClient.put<Vehicle>(`/vehicles/${id}`, vehicle);
    return data;
  },

  async toggleAvailability(id: string): Promise<Vehicle> {
    const { data } = await apiClient.patch<Vehicle>(`/vehicles/${id}/toggle`);
    return data;
  },

  async updateStatus(id: string, rentalStatus: VehicleRentalStatus): Promise<Vehicle> {
    const { data } = await apiClient.patch<Vehicle>(`/vehicles/${id}/status`, { rentalStatus });
    return data;
  },

  async delete(id: string): Promise<void> {
    await apiClient.delete(`/vehicles/${id}`);
  },
};

export default vehicleService;

File: backend/src/main/java/com/voyago/service/VehicleService.java
package com.voyago.service;

import com.voyago.model.Vehicle;
import com.voyago.repository.VehicleRepository;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Service;

import java.util.List;

@Service
@RequiredArgsConstructor
public class VehicleService {

    private final VehicleRepository vehicleRepository;

    public List<Vehicle> getAll() {
        return vehicleRepository.findAll();
    }

    public Vehicle getById(String id) {
        return vehicleRepository.findById(id)
                .orElseThrow(() -> new RuntimeException("Vehicle not found: " + id));
    }

    public List<Vehicle> getByDestination(String destinationId) {
        return vehicleRepository.findByDestinationId(destinationId);
    }

    public List<Vehicle> getAvailable() {
        return vehicleRepository.findByIsAvailableTrue();
    }

    public Vehicle create(Vehicle vehicle) {
        return vehicleRepository.save(vehicle);
    }

    public Vehicle update(String id, Vehicle updated) {
        Vehicle existing = getById(id);
        updated.setId(existing.getId());
        return vehicleRepository.save(updated);
    }

    public Vehicle toggleAvailability(String id) {
        Vehicle vehicle = getById(id);
        vehicle.setAvailable(!vehicle.isAvailable());
        vehicle.setRentalStatus(vehicle.isAvailable() ? "AVAILABLE" : "RENTED");
        return vehicleRepository.save(vehicle);
    }

    public void delete(String id) {
        vehicleRepository.deleteById(id);
    }
}
